File: packages/core/src/capture/validator.ts
import type { KeystrokeEvent } from "../types/events";

/**
 * Validate a keystroke event against sequence/timestamp ordering.
 * Throws on invalid event.
 */
export function validateEvent(
  event: KeystrokeEvent,
  lastSequence: number,
  lastTimestamp: number,
): void {
  if (event.sequence_number <= lastSequence && lastSequence > 0) {
    throw new Error(
      `Sequence regression: got ${event.sequence_number} expected > ${lastSequence}`,
    );
  }

  if (event.timestamp < lastTimestamp - 1) {
    throw new Error(
      `Timestamp regression: got ${event.timestamp} expected >= ${lastTimestamp}`,
    );
  }

  if (event.key === "" && event.code === "") {
    throw new Error("Empty key and code");
  }
}

File: apps/desktop/src-tauri/src/capture/validator.rs
use super::events::KeystrokeEvent;

pub fn validate_event(event: &KeystrokeEvent, last_sequence: u64, last_timestamp: f64) -> Result<(), String> {
    if event.sequence_number <= last_sequence && last_sequence > 0 {
        return Err(format!(
            "Sequence regression: got {} expected > {}",
            event.sequence_number, last_sequence
        ));
    }

    if event.timestamp < last_timestamp - 1.0 {
        return Err(format!(
            "Timestamp regression: got {} expected >= {}",
            event.timestamp, last_timestamp
        ));
    }

    if event.key.is_empty() && event.code.is_empty() {
        return Err("Empty key and code".to_string());
    }

    Ok(())
}
